src/daft-micropartition/src/micropartition.rs

```diff
@@ -693,6 +700,68 @@ pub(crate) fn read_json_into_micropartition(
     }
 }
 
+#[allow(clippy::too_many_arguments)]
+fn _read_parquet_into_loaded_micropartition(
+    io_client: Arc<IOClient>,
+    runtime_handle: Arc<tokio::runtime::Runtime>,
+    uris: &[&str],
+    columns: Option<&[&str]>,
+    start_offset: Option<usize>,
+    num_rows: Option<usize>,
+    row_groups: Option<Vec<Option<Vec<i64>>>>,
+    predicate: Option<ExprRef>,
+    partition_spec: Option<&PartitionSpec>,
+    io_stats: Option<IOStatsRef>,
+    num_parallel_tasks: usize,
+    schema_infer_options: &ParquetSchemaInferenceOptions,
+    catalog_provided_schema: Option<SchemaRef>,
+) -> DaftResult<MicroPartition> {
+    let all_tables = read_parquet_bulk(
+        uris,
+        columns,
+        start_offset,
+        num_rows,
+        row_groups,
+        predicate,
+        io_client,
+        io_stats,
+        num_parallel_tasks,
+        runtime_handle,
+        schema_infer_options,
+    )?;
+
+    // Prefer using the `catalog_provided_schema` but fall back onto inferred schema from Parquet files
+    let full_daft_schema = match catalog_provided_schema {
+        Some(catalog_provided_schema) => catalog_provided_schema,
+        None => {
+            let unioned_schema = all_tables
+                .iter()
+                .map(|t| t.schema.clone())
+                .try_reduce(|l, r| DaftResult::Ok(l.union(&r)?.into()))?;
+            unioned_schema.expect("we need at least 1 schema")
+        }
+    };
+
+    // Hack to avoid to owned schema
+    let full_daft_schema = Schema {
+        fields: full_daft_schema.fields.clone(),
+    };
+    let pruned_daft_schema = prune_fields_from_schema(full_daft_schema, columns)?;
+
+    let fill_map = partition_spec.map(|pspec| pspec.to_fill_map());
+    let all_tables = all_tables
+        .into_iter()
+        .map(|t| t.cast_to_schema_with_fill(&pruned_daft_schema, fill_map.as_ref()))
+        .collect::<DaftResult<Vec<_>>>()?;
+
+    // TODO: we can pass in stats here to optimize downstream workloads such as join
```

Comment:
When we eventually do this, we'll need to be sure to cast these stats to the `pruned_daft_schema` with the `fill_map` as well!

Reply:
That's correct, I'll add to the comment as well!